src/main/scala/com/snowplowanalytics/snowplow/postgres/loader/streaming/source.scala

```diff
@@ -0,0 +1,150 @@
+                                                      config: LoaderConfig.Source) =
+    config match {
+      case LoaderConfig.Source.Kinesis(appName, streamName, region, position) =>
+        KinesisConsumerSettings.apply(streamName, appName, region, initialPositionInStream = position.unwrap) match {
+          case Right(settings) =>
+            readFromKinesisStream[F](settings).map(processRecord(purpose)).asRight
+          case Left(error) =>
+            error.asLeft
+        }
+      case LoaderConfig.Source.PubSub(projectId, subscriptionId) =>
+        implicit val decoder: MessageDecoder[Either[BadData, Data]] = pubsubDataDecoder(purpose)
+        val project = ProjectId(projectId)
+        val subscription = Subscription(subscriptionId)
+        val pubsubConfig = PubsubGoogleConsumerConfig[F](onFailedTerminate = pubsubOnFailedTerminate[F])
+        PubsubGoogleConsumer.subscribeAndAck[F, Either[BadData, Data]](blocker, project, subscription, pubsubErrorHandler[F], pubsubConfig).asRight
+    }
+
+  /**
+   * Parse Kinesis record into a valid Loader's record, either enriched event or self-describing JSON,
+   * depending on purpose of the Loader
+   */
+  def processRecord(kind: Purpose)(record: CommittableRecord): Either[BadData, Data] = {
+    val string = try {
+      StandardCharsets.UTF_8.decode(record.record.data()).toString.asRight[BadData]
+    } catch {
+      case _: IllegalArgumentException =>
+        val payload = StandardCharsets.UTF_8.decode(Base64.getEncoder.encode(record.record.data())).toString
+        kind match {
+          case Purpose.Enriched =>
+            val badRow = BadRow.LoaderParsingError(Cli.processor, NotTSV, Payload.RawPayload(payload))
+            BadData.BadEnriched(badRow).asLeft
+          case Purpose.SelfDescribing =>
+            BadData.BadJson(payload, "Cannot deserialize self-describing JSON from Kinesis record").asLeft
+        }
+    }
+
+    string.flatMap { payload =>
+      kind match {
+        case Purpose.Enriched =>
+          parseEventString(payload).map(Data.Snowplow.apply)
+        case Purpose.SelfDescribing =>
+          parseJson(payload).map(Data.SelfDescribing.apply)
+      }
+    }
+  }
+
+  def parseEventString(s: String): Either[BadData, Event] =
+    Event.parse(s).toEither.leftMap { error =>
+      val badRow = BadRow.LoaderParsingError(Cli.processor, error, Payload.RawPayload(s))
+      BadData.BadEnriched(badRow)
+    }
+
+  def parseJson(s: String): Either[BadData, SelfDescribingData[Json]] =
+    parseCirce(s)
+      .leftMap(_.show)
+      .flatMap(json => SelfDescribingData.parse[Json](json).leftMap(_.message(json.noSpaces)))
+      .leftMap(error => BadData.BadJson(s, error))
+
+  /** Kind of data flowing through the Loader */
+  sealed trait Data extends Product with Serializable {
+    def snowplow: Boolean = this match {
+      case _: Data.Snowplow => true
+      case _: Data.SelfDescribing => false
+    }
+  }
```

Comment:
Not exactly. `Purpose` is a value by itself, whereas `snowplow` is a property of datum. It could return `Purpose` with the same effect, but semantically `Purpose` is a property of config/loader instance, not datum.